presto-parquet/src/main/java/io/prestosql/parquet/reader/ShortDecimalColumnReader.java

```diff
@@ -47,54 +52,97 @@
+                value = getShortDecimalValue(valuesReader.readBytes().getBytes());
             }
 
-            if (isShortDecimal(prestoType)) {
-                validateDecimal((DecimalType) prestoType);
-                // TODO: validate that value fits Presto decimal precision
-                prestoType.writeLong(blockBuilder, decimalValue);
-            }
-            else if (isLongDecimal(prestoType)) {
-                validateDecimal((DecimalType) prestoType);
-                // TODO: validate that value fits Presto decimal precision
-                prestoType.writeSlice(blockBuilder, unscaledDecimal(decimalValue));
-            }
-            else if (prestoType.equals(TINYINT) || prestoType.equals(SMALLINT) || prestoType.equals(INTEGER) || prestoType.equals(BIGINT)) {
-                if (parquetDecimalType.getScale() != 0) {
-                    throw new ParquetDecodingException(format(
-                            "Parquet decimal column type with non-zero scale (%s) cannot be converted to Presto %s column type",
+            if (prestoType instanceof DecimalType) {
+                DecimalType prestoDecimalType = (DecimalType) prestoType;
+
+                if (isShortDecimal(prestoDecimalType)) {
+                    long rescale = longTenToNth(Math.abs(prestoDecimalType.getScale() - parquetDecimalType.getScale()));
+                    long convertedValue = shortToShortCast(
+                            value,
+                            parquetDecimalType.getPrecision(),
+                            parquetDecimalType.getScale(),
+                            prestoDecimalType.getPrecision(),
+                            prestoDecimalType.getScale(),
+                            rescale,
+                            rescale / 2);
+
+                    prestoType.writeLong(blockBuilder, convertedValue);
+                }
+                else if (isLongDecimal(prestoDecimalType)) {
+                    prestoType.writeSlice(blockBuilder, shortToLongCast(
+                            value,
+                            parquetDecimalType.getPrecision(),
                             parquetDecimalType.getScale(),
-                            prestoType));
+                            prestoDecimalType.getPrecision(),
+                            prestoDecimalType.getScale()));
                 }
-                prestoType.writeLong(blockBuilder, decimalValue);
             }
             else {
-                throw new ParquetDecodingException(format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
+                if (parquetDecimalType.getScale() != 0) {
+                    throw new PrestoException(NOT_SUPPORTED, format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
+                }
+
+                long rescale = longTenToNth(parquetDecimalType.getScale());
+                long convertedValue = shortToShortCast(
```

Comment:
I realized this `shortToShortCast` here is a no-op, because we know `parquetDecimalType.getScale() == 0`